components/layout_2020/display_list/mod.rs

```diff
@@ -199,38 +206,51 @@ impl<'a> BuilderForBoxFragment<'a> {
         })
     }
 
-    fn with_border_edge_clip(
-        &mut self,
-        builder: &mut DisplayListBuilder,
-        common: &mut wr::CommonItemProperties,
-    ) {
-        let initialized = self.border_edge_clip_id.init_once(|| {
-            if self.border_radius.is_zero() {
-                None
-            } else {
-                Some(builder.wr.define_clip(
-                    &builder.current_space_and_clip,
-                    self.border_rect,
-                    Some(wr::ComplexClipRegion {
-                        rect: self.border_rect,
-                        radii: self.border_radius,
-                        mode: wr::ClipMode::Clip,
-                    }),
-                    None,
-                ))
-            }
-        });
-        if let Some(clip_id) = *initialized {
-            common.clip_id = clip_id
-        }
+    fn border_edge_clip(&self, builder: &mut DisplayListBuilder) -> Option<wr::ClipId> {
+        *self
+            .border_edge_clip_id
+            .init_once(|| clip_for_radii(self.border_radius, self.border_rect, builder))
+    }
+
+    fn padding_edge_clip(&self, builder: &mut DisplayListBuilder) -> Option<wr::ClipId> {
+        *self.padding_edge_clip_id.init_once(|| {
+            clip_for_radii(
+                inner_radii(
+                    self.border_radius,
+                    self.fragment
+                        .border
+                        .to_physical(self.fragment.style.writing_mode)
+                        .to_webrender(),
+                ),
+                self.border_rect,
+                builder,
+            )
+        })
+    }
+
+    fn content_edge_clip(&self, builder: &mut DisplayListBuilder) -> Option<wr::ClipId> {
+        *self.content_edge_clip_id.init_once(|| {
+            clip_for_radii(
+                inner_radii(
+                    self.border_radius,
+                    (&self.fragment.border + &self.fragment.padding)
+                        .to_physical(self.fragment.style.writing_mode)
+                        .to_webrender(),
+                ),
+                self.border_rect,
+                builder,
+            )
+        })
     }
 
     fn build(&mut self, builder: &mut DisplayListBuilder) {
         let hit_info = hit_info(&self.fragment.style, self.fragment.tag, Cursor::Default);
         if hit_info.is_some() {
             let mut common = builder.common_properties(self.border_rect);
             common.hit_info = hit_info;
-            self.with_border_edge_clip(builder, &mut common);
+            if let Some(clip_id) = self.border_edge_clip(builder) {
```

Comment:
At some point it would be good to keep a stack of clips and spatial node ids, but for the moment I don't think it's necessary.

Reply:
Yes, having that stack on the call stack is the idea of:

https://github.com/servo/servo/blob/12693b51f510761d5dd58a9103596a192edb803c/components/layout_2020/display_list.rs#L63-L70

… which I copied from Layout 2013 code, but ended up not using for backgrounds because each background layer has a potentially-different `background-clip` value that only applies to one display item, so pushing/popping to the stack isn’t as useful.